src/include/storage/index/index_populator.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+
+#include <vector>
+#include "loggers/main_logger.h"
+#include "storage/index/index.h"
+#include "storage/sql_table.h"
+#include "transaction/transaction_context.h"
+
+namespace terrier::storage::index {
+
+/**
+ * An index populator is a helper class that populate the latest version of all tuples in the table
+ * visible to the current transaction, pack target columns into ProjectedRow's and insert into
+ * the given index.
+ *
+ * This class can only be used when creating an index.
+ */
+class IndexPopulator {
+ public:
+  /**
+   * The method populates all tuples in the table visible to current transaction with latest version,
+   * then pack target columns into ProjectedRow's and insert into the given index.
+   *
+   * @param txn current transaction context
+   * @param sql_table the target sql table
+   * @param index target index inserted into
+   */
+  static void PopulateIndex(transaction::TransactionContext *txn,                                         // NOLINT
+                            SqlTable &sql_table, const IndexKeySchema &index_key_schema, Index &index) {  // NOLINT
+    // Create the projected row for the index
+    const IndexMetadata &metadata = index.GetIndexMetadata();
+    const auto &index_pr_init = metadata.GetProjectedRowInitializer();
+    auto *index_pr_buf = common::AllocationUtil::AllocateAligned(index_pr_init.ProjectedRowSize());
+    ProjectedRow *indkey_pr = index_pr_init.InitializeRow(index_pr_buf);
+
+    // Create the projected row for the sql table
+    std::vector<catalog::col_oid_t> col_oids;
+    for (const auto &it : index_key_schema) {
+      col_oids.emplace_back(catalog::col_oid_t(!it.GetOid()));
+    }
+    auto table_pr_init = sql_table.InitializerForProjectedRow(col_oids).first;
+    auto *table_pr_buf = common::AllocationUtil::AllocateAligned(table_pr_init.ProjectedRowSize());
+    ProjectedRow *select_pr = table_pr_init.InitializeRow(table_pr_buf);
+
+    // Record the col_id of each column
+    const col_id_t *columns = select_pr->ColumnIds();
+    uint16_t num_cols = select_pr->NumColumns();
+    std::vector<col_id_t> sql_table_cols(columns, columns + num_cols);
+
+    for (const auto &it : sql_table) {
```

Comment:
Are you implementing the blocking approach here? If so, should it acquire lock for the sql table before populating the index?

Reply:
Yes, you are right. Actually, here we just write some necessary methods here and we will write a wrapper function later to show steps of blocking approach. That wrapper function will include acquiring and releasing locks.